### PR TITLE
optimize the optimistic unchoke logic-1.1

### PR DESCRIPTION
**Pull request by arvidn**
and reduce pointer indirection in optimistic unchoke logic

### CODE REVIEW

**Comment by ssiloti** [comment]
This kills the reputation plugin :(

**Comment by arvidn** [comment]
oh.. it's using the optimistic unchoke? is that the only reward one gets for having good reputation?
it was hard for me to see all the costs associated with wrapping and unwrapping all those pointers.

what if there was a way to set the optimistic unchoke priority instead? or maybe just set the time of last optimistic unchoke (and if set to a long time ago, it will have high priority)

**Comment by ssiloti** [comment]
Yes and yes, optimistic unchoking is currently the only way the reputation plugin has to reward good reputation. Internally the reputation plugin stores peer reputation as a double (larger is better) so if there was a way to set a peer's priority via peer_connection_handle (probably as a uint64) you could then use that when sorting the opt_unchoke vector. This would remove the pointer indirection at the cost of less flexibility for plugins to control the optimistic unchoke order.

**Comment by arvidn** [comment]
ok. looking at your code. you have a double indicating the optimistic unchoke priority that you want to sort by. I would imagine that your use case would benefit (performance vice) to not build that temporary vector, sort it and put it back in the one provided by libtorrent too.

The question then is, how do you map a double to a last-optimistic-unchoke timestamp? I'm not sure, I suppose it depends. Are there bounds on the score one can have?

in the timestamps-space, lower values are higher priority, and in the score-space, higher values are higher priority. Perhaps we could do something like:

peer->set_last_optimistic_unchoke(now() - milliseconds(score * 1000.0));

This is assuming score is positive and that when interpreting it as seconds, millisecond resolution is good enough. It also assumes that the current time is not very significant, assuming scores are updated at different times for different peers, "now" will be different and affect the priority. (by increasing the score-factor, the current time can be made less relevant).

However, one benefit of including the current time is to prevent starvation. If the optimistic unchoke priority is strictly weighed by the score, a low-score peer may never be optimistically unchoked, and can might never get an opportunity to prove its worth and get good reputation.

This last point is a bit subtle but potentially quite important. Did you consider affecting the regular unchoke logic instead?

Alternatively, one can think of it as:

optimistic unchoke logic is for finding new peers that are good, to give them an opportunity to reciprocate. Instead of a blind, random round-robin search, weigh the search towards the ones with good reputation first, because they are more likely to provide good return on investment early and accelerate finding the optimal peer set.

With that reasoning, it would primarily make sense for the score to affect the initial order of optimistic unchoking, but still have a strict round-robin after that.

**Comment by arvidn** [comment]
I'm also hoping to make the libtorrent implementation for optimistic unchoking a lot more efficient, by making a single pass over the peers. I believe this patch would also allow a future where optimistic and regular unchokes are done per torrent, instead of per session. The benefit of that is parallelism, torrents can be less dependent on each other and potentially run in separate threads.

**Comment by ssiloti** [comment]
There are no bounds on the reputation value. In fact the use of double is an artifact of the original ratio based policy. With the current policy the reputation is derived from the peer's "balance of transfer", i.e. the number of bytes the client/intermediaries have downloaded from the peer minus the bytes they have uploaded to it. It should probably be changed to be a uint64.

Because of this mapping to a timestamp is not trivial, without some sort of normalization the reputation will likely completely swamp the actual passage of time, or vice-versa depending on the constant factors used.

Using reputation to influence regular unchoking is troublesome for much the same reason, some balance would have to be struck between the existing rate-based logic and reputation. It's also not clear to me that it makes sense to allow reputation to override the rate-based unchocking logic, a peer that the client is receiving data from at a high rate right now is a much clearer signal of "goodness" than reputation.

Providing a one-time priority boost for optimistically unchocking a peer when it first connects might make sense, although I'm kind of worried that might be too weak of an influence to provide a noticeable reward for good reputation.

**Comment by ssiloti** [comment]
After looking at the code I think using reputation as a tie breaker when two peers have the same last_optimistically_unchoked time might work well enough. This would typically only come into play between peers which have never been optimistically unchoked (both would have last_optimistically_unchoked == 0) so it would effectively create a one-time boost. This gives reputation a pretty limited influence over unchocking, but I'm fine with starting off conservative. If it turns out to be insufficient there's no harm done and we can try incrementally stronger policies.

**Comment by arvidn** [comment]
once you're a seed, all regular unchoking could probably be controlled entirely by reputation, as a reward mechanism.

File: src/session_impl.cpp [hunk 1]
```diff
@@ -3784,12 +3786,14 @@ namespace aux {
...
 			torrent* t = p->associated_torrent().lock().get();
 			if (!t) continue;
+			// TODO: 3 peers should know whether their torrent is paused or not,
+			// instead of having to ask it over and over again
 			if (t->is_paused()) continue;
 
 			if (pi->optimistically_unchoked)
 			{
 				TORRENT_ASSERT(!p->is_choked());
-				opt_unchoke.push_back(peer_connection_handle(*i));
```
Comment (arvidn) [inline src/session_impl.cpp]: it looks suspicious that the same peer can be added twice here..

**Comment by ssiloti** [comment]
Yes, reputation could be used for regular unchoking while seeding, but the logic would need to be very carefully written. If unchoking were driven entirely by reputation the client could end up severely under-utilizing its upload capacity.

File: src/session_impl.cpp (hunk 1 above)
Comment (ssiloti) [inline src/session_impl.cpp]: I think the intent is that optimistically_unchoked implies the peer is unchoked so if it is true then the second condition will never be true due to the is_choked() term.

**Comment by arvidn** [comment]
the resolution of the last_optimistically_unchoked is only 16 bits. It's initialized to 0, which represents something between 9 and 18 hours ago. It would be easy to just look for peers with zeroes and initialize them to something based on their score, to just create an initial bias of which of the never-optimistically unchoked to optimistically unchoke next. I think the main challenge is to turn an unbounded double into a 16 bit score. realistically it's probably just 15 bits, since if you get too high timestamps, you may start interfere with the peers that have been optimistically unchoked for real once (but maybe that's not a problem, if a peer has that low reputation)

**Comment by arvidn** [comment]
I will add a simulation exercising the optimistic unchoke logic too. I wonder if perhaps this patch is snow-balling a bit too, and perhaps should go into master.

**Comment by ssiloti** [comment]
To make that work I'd need to keep track of the highest reputation of all active sessions. Then I could normalize the reputations so that the highest reputation gives an initial timestamp of zero and a reputation of zero gives some arbitrary high initial timestamp, say 2^14. That would actually be pretty nasty because the normalization factor could change radically at any moment if a very high reputation session is opened. This would compromise all existing normalized initial values. I supposed I could normalize to the highest reputation of any peer in the database, but that could cause problems if there is a very high reputation peer which causes all active peer's initial timestamps to be packed near the max initial timestamp. I could try to prevent that by using a logarithmic scale to normalize, ugh...

Adding an int64_t to torrent_peer and using that as a tie breaker when last_optimistically_unchoked is equal is radically simpler.

**Comment by ssiloti** [comment]
Thinking about it a little more, keeping track of the highest reputation in the database isn't even practical because it could change based on the ever shifting reputations of intermediaries. I'd have to use some approximation kludge.

**Comment by arvidn** [comment]
ok. I think it's probably best to revert back to the original mechanism. I'm thinking of perhaps extensions can express that they overload certain functions, and have those called, rather than always call all extensions.

In this case the logic could be a lot simpler/efficient if no extension implemented custom optimistic unchoke logic for instance.

maybe the long term way to introduce customization points would be to subscribe to events, and have callbacks be called. that way, libtorrent doesn't have to call functions on every extension just in case they customize it.

**Comment by ssiloti** [comment]
Hmm, mayyybe I could make this work if I set the normalization max to double the highest active session's reputation every a session's reputation exceeded the existing max. That way I wouldn't have to recompute the biases too often. I'd still need to use a logarithmic scale, but that's not too big of a deal.

**Comment by ssiloti** [comment]
Looking at the possibility of integrating reputation into the seed chocking algorithms, I think it's very doable and desirable. It obviously doesn't interact well with the use of last_optimistically_unchoked to signal reputation though. Maybe instead of calling into each extension on every optimistic unchoke we could add a way for extensions to register a comparison function that takes two peer_connection_handles. We could store it as a function pointer+void pointer pair in session_impl and call it whenever the internal comparison function produces a tie.

**Comment by ssiloti** [comment]
Hmm, that would require opt_unchoke go back to being a vector of peer_connection_handles. I could instead make it a vector of <torrent_peer*, uint64_t> pairs where I call a get_unchoke_priority function to get the priority for peers with a last_optimistically_unchoked of zero. Arvid, what's your preference?

**Comment by arvidn** [comment]
I'm sort of thinking of keeping the plugin interface more or less the same for now. I think there probably needs to be some more thinking about how this could be made cheaper, more flexible and simpler.

The last patch restores the plugin interface except it adds a new function for indicating interest in having the optimistic unchoke callback be called. if no session plugin is interested, the conversions will be skipped.

**Comment by arvidn** [comment]
what I mean is that it's probably not appropriate to make too big changes for 1.1

**Comment by ssiloti** [comment]
One way I was thinking of to get finer grained information about what a plugin actually implements was to break the plugin interfaces up into smaller abstract classes, at least for the functions that get called a lot. Then instead of storing a single monolithic vector of pointers to extensions you have a handful of vectors, one for each abstract class. When a plugin is added you attempt to dynamic_cast the plugin pointer to the different classes and add it to the vector if the cast succeeds.

This way you make the feature checks up-front instead of checking at every call.

**Comment by arvidn** [comment]
yeah, I was thinking of multiple extension interfaces too. the dynamic_cast is a neat way of implicitly declare what's implemented. I was also thinking along the line of signal-slots, where an extension can set a std::function to be called, rather than the callbacks being virtual functions. with lambdas I think that could be made fairly simple to use too.

**Comment by ssiloti** [comment]
I was bored so I implemented a proof-of-concept for fine-grained interfaces

https://github.com/ssiloti/libtorrent/commit/d3d30ef59dcafc2ed05b2e9a2fd2db4eff9957c9

**Comment by LordNyriox** [comment]
@arvidn Might you merge this or @ssiloti's patch at some point? 

Thanks, 
~ Nyriox

**Pull request merged**